Add FinOps Best Practices / ARG resource library (#908)
microsoft/finops-toolkit: docs/_docs/bestpractices/networking/networking.md

---
layout: default
parent: Best practices

> permalink: /bestpractices/networking

Reviewer: ```suggestion
permalink: /best-practices/networking
```

nav_order: 2
title: Networking
author: arclares
ms.date: 08/16/2024
ms.service: finops
description: 'Discover essential FinOps best practices to optimize cost efficiency and governance for your Azure resources.'

---

<span class="fs-9 d-block mb-4">Networking</span>

> Discover essential FinOps best practices to optimize cost efficiency and governance for your Azure networking resources.

Reviewer: nit: All of these are the same basic text. Might be nice to make them a little unique. This can happen later. Just a thought.

{: .fs-6 .fw-300 }

[Share feedback](#️-looking-for-more){: .btn .fs-5 .mb-4 .mb-md-0 .mr-4 }

<details open markdown="1">
   <summary class="fs-2 text-uppercase">On this page</summary>

- [🛡️ Azure Firewall](#azure-firewall)
- [🔗  Load Balancer](#load-balancer)
- [🌐 Public IP Address](#public-ip-address)
- [🖧 Virtual Network Gateway](#virtual-network-gateway)
- [📡 Express Route](express-route)
- [📶  Private DNS](#private-dns)
- [🙋‍♀️ Looking for more?](#️-looking-for-more)
- [🧰 Related tools](#-related-tools)

</details>

---

> ## Azure firewall 

Reviewer: Casing is confusing, I know... 😕
```suggestion
## Azure Firewall 
```

#### Category

Optimization

#### Query

<details>
  <summary>Click to view the code</summary>
  <div class="code-block">
    <pre><code> resources 
  | where type =~ 'Microsoft.Network/azureFirewalls' and properties.sku.tier=="Premium"
  | project FWID=id, firewallName=name, SkuTier=tostring(properties.sku.tier), resourceGroup, location
  | join kind=inner (
      resources
      | where type =~ 'microsoft.network/firewallpolicies'
      | mv-expand properties.firewalls
      | extend intrusionDetection=tostring(properties.intrusionDetection contains "Alert" or properties.intrusionDetection contains "Deny"), transportSecurity=tostring(properties.transportSecurity contains "keyVaultSecretId")
      | extend FWID=tostring(properties_firewalls.id)
      | where intrusionDetection == "False" and transportSecurity == "False"
      | project PolicyName=name, PolicySKU=tostring(properties.sku.tier), intrusionDetection, transportSecurity, FWID
  ) on FWID</code></pre>
  </div>
</details>

> ### Query: Azure firewall and associated subnets analysis

Reviewer: ```suggestion
### Query: Azure Firewall and associated subnets analysis
```

#### Category

Optimization

#### Query

<details>
  <summary>Click to view the code</summary>
  <div class="code-block">
    <pre><code> resources 
  | where type =~ 'Microsoft.Network/azureFirewalls' and properties.sku.tier=="Premium"
  | project FWID=id, firewallName=name, SkuTier=tostring(properties.sku.tier), resourceGroup, location
  | join kind=inner (
      resources
      | where type =~ 'microsoft.network/firewallpolicies'
      | mv-expand properties.firewalls
      | extend intrusionDetection=tostring(properties.intrusionDetection contains "Alert" or properties.intrusionDetection contains "Deny"), transportSecurity=tostring(properties.transportSecurity contains "keyVaultSecretId")
      | extend FWID=tostring(properties_firewalls.id)
      | where intrusionDetection == "False" and transportSecurity == "False"
      | project PolicyName=name, PolicySKU=tostring(properties.sku.tier), intrusionDetection, transportSecurity, FWID
  ) on FWID</code></pre>
  </div>
</details>

> ## Application gateway

Reviewer: ```suggestion
<br>

## Application Gateway
```

| where type =~ 'Microsoft.Network/applicationGateways'
| extend backendPoolsCount = array_length(properties.backendAddressPools),SKUName= tostring(properties.sku.name), SKUTier= tostring(properties.sku.tier),SKUCapacity=properties.sku.capacity,backendPools=properties.backendAddressPools,resourceGroup=strcat('/subscriptions/',subscriptionId,'/resourceGroups/',resourceGroup)
| project id, name, SKUName, SKUTier, SKUCapacity,resourceGroup,subscriptionId
| join (
    resources
    | where type =~ 'Microsoft.Network/applicationGateways'
    | mvexpand backendPools = properties.backendAddressPools
    | extend backendIPCount = array_length(backendPools.properties.backendIPConfigurations)
    | extend backendAddressesCount = array_length(backendPools.properties.backendAddresses)
    | extend backendPoolName  = backendPools.properties.backendAddressPools.name
    | summarize backendIPCount = sum(backendIPCount) ,backendAddressesCount=sum(backendAddressesCount) by id
) on id
| project-away id1
| where  (backendIPCount == 0 or isempty(backendIPCount)) and (backendAddressesCount==0 or isempty(backendAddressesCount))
| order by id asc
</code></pre>
  </div>
</details>

> ## Load Balancer

Reviewer: ```suggestion
<br>

## Load Balancer
```

### Query: Idle load balancers

This Azure Resource Graph (ARG) query analyzes Azure load balancers and their associated backend pools within your Azure environment. It provides insights into which load balancers have empty backend pools, indicating they may be idle and potentially unnecessary.

#### Category

Optimization

#### Query

<details>
  <summary>Click to view the code</summary>
  <div class="code-block">
    <pre><code> resources 
    | extend resourceGroup=strcat('/subscriptions/',subscriptionId,'/resourceGroups/',resourceGroup)
    | extend SKUName=tostring(sku.name)
    | extend SKUTier=tostring(sku.tier)
    | extend location,backendAddressPools = properties.backendAddressPools
    | where type =~ 'microsoft.network/loadbalancers' and array_length(backendAddressPools) == 0 and sku.name!='Basic'
    | order by id asc
| project id,name, SKUName,SKUTier,backendAddressPools, location,resourceGroup, subscriptionId
</code></pre>
  </div>
</details>

> ## Public IP address

Reviewer: ```suggestion
<br>

## Public IP address
```

| project PublicIpId,IPName, RoutingMethod,SKUName, resourceGroup, Location, AllocationMethod, subscriptionId
</code></pre>
  </div>
</details>

### Query: Check public IP addresses' DDoS protection policy

If you need to protect fewer than 15 public IP resources, the IP protection tier is the more cost-effective option. However, if you have more than 15 public IP resources to protect, then the network protection tier becomes more cost-effective.

#### Category

Optimization

#### Query

<details>
  <summary>Click to view the code</summary>
  <div class="code-block">
    <pre><code> resources
| where type == "microsoft.network/publicipaddresses"
| project ddosProtection=tostring(properties.ddosSettings), name
| where ddosProtection has "Enabled"
| count
| project TotalIpsProtected = Count
| extend CheckIpsProtected = iff(TotalIpsProtected >= 15,"Enable Network Protection tier", "Enable PIP DDoS Protection")
</code></pre>
  </div>
</details>

> ## Virtual Network Gateway

Reviewer: ```suggestion
<br>

## Virtual Network Gateway
```

| extend resourceGroup =strcat('/subscriptions/',subscriptionId,'/resourceGroups/',resourceGroup)
| project id, GWName=name,resourceGroup,location,subscriptionId
| join kind = leftouter(
    resources
    | where type == "microsoft.network/connections"
    | extend id = tostring(properties.virtualNetworkGateway1.id)
    | project id
) on id
| where isempty(id1)
| project id, GWName,resourceGroup,location,subscriptionId,status=id
</code></pre>
  </div>
</details>

### Query: Check for idle NAT gateway

This Azure Resource Graph (ARG) query analyzes NAT gateways within your Azure environment to identify those that are idle.

#### Category

Optimization

#### Query

<details>
  <summary>Click to view the code</summary>
  <div class="code-block">
    <pre><code> resources
| where type == "microsoft.network/natgateways" and isnull(properties.subnets)
| project id, GWName=name, SKUName=tostring(sku.name), SKUTier=tostring(sku.tier), Location=location ,resourceGroup=tostring(strcat('/subscriptions/',subscriptionId,'/resourceGroups/',resourceGroup)),subnet=tostring(properties.subnet), subscriptionId
</code></pre>
  </div>
</details>

> ## Express Route

Reviewer: ```suggestion
<br>

## ExpressRoute
```

> ### Query: Idle Express Route

Reviewer: ```suggestion
### Query: Idle ExpressRoute circuits
```

> This Azure Resource Graph (ARG) query analyzes Express Route circuits within your Azure environment to identify those without a completed circuit.

Reviewer: ```suggestion
This Azure Resource Graph (ARG) query analyzes ExpressRoute circuits within your Azure environment to identify those without a completed circuit.
```

#### Category

Optimization

#### Query

<details>
  <summary>Click to view the code</summary>
  <div class="code-block">
    <pre><code> resources
| where type =~ 'Microsoft.Network/expressRouteCircuits' and properties.serviceProviderProvisioningState == "NotProvisioned"
| extend ServiceLocation=tostring(properties.serviceProviderProperties.peeringLocation), ServiceProvider=tostring(properties.serviceProviderProperties.serviceProviderName), BandwidthInMbps=tostring(properties.serviceProviderProperties.bandwidthInMbps)
| project   ERId=id,ERName = name, ERRG = resourceGroup, SKUName=tostring(sku.name), SKUTier=tostring(sku.tier), SKUFamily=tostring(sku.family), ERLocation = location, ServiceLocation, ServiceProvider, BandwidthInMbps

</code></pre>
  </div>
</details>

> ## Private DNS

Reviewer: ```suggestion
<br>

## Private DNS
```

### Query: Private DNS

This Azure Resource Graph (ARG) query analyzes Private DNS zones within your Azure environment to identify those without Virtual Network Links.

#### Category

Optimization

#### Query

<details>
  <summary>Click to view the code</summary>
  <div class="code-block">
    <pre><code> resources
| where type == "microsoft.network/privatednszones" and properties.numberOfVirtualNetworkLinks == 0
| project id, PrivateDNSName=name, NumberOfRecordSets=tostring(properties.numberOfRecordSets),resourceGroup=tostring(strcat('/subscriptions/',subscriptionId,'/resourceGroups/',resourceGroup)),vNets=tostring(properties.properties.numberOfVirtualNetworkLinks), subscriptionId
</code></pre>
  </div>
</details>

> ## 🙋‍♀️ Looking for more?

Reviewer: ```suggestion
<br>

## 🙋‍♀️ Looking for more?
```
